eitprocessing/datahandling/intervaldata.py

```diff
@@ -0,0 +1,185 @@
+        values: an optional list of values with the same length as time_ranges
+        parameters: parameters used to derive the data
+        derived_from: list of data sets this data was derived from
+        description: extended human readible description of the data
+        partial_inclusion: whether to include a trimmed version of a time range when selecting data
+    """
+
+    label: str
+    name: str
+    unit: str | None
+    category: str
+    time_ranges: list[TimeRange | tuple[float, float]]
+    values: list[Any] | None = None
+    parameters: dict[str, Any] = field(default_factory=dict)
+    derived_from: list[Any] = field(default_factory=list)
+    description: str = ""
+    partial_inclusion: bool = False
+
+    def __post_init__(self) -> None:
+        self.time_ranges = [TimeRange._make(time_range) for time_range in self.time_ranges]
+        self._check_equivalence = ["unit", "category", "parameters", "partial_inclusion"]
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}('{self.label}')"
+
+    def __len__(self) -> int:
+        return len(self.time_ranges)
+
+    def select_by_time(  # noqa: C901
+        self,
+        start_time: float | None = None,
+        end_time: float | None = None,
+        partial_inclusion: bool | None = None,
+        newlabel: str | None = None,
+    ) -> Self:
+        """Return only period data that overlaps (partly) with start and end time.
+
+        Other types of data (e.g. ContinuousData and SparseData) support the start_inclusive and end_inclusive
+        arguments. PeriodData does not. That means that selection by time of PeriodData probably works slightly
+        different than other types of selecting/slicing data.
+        """
+        if partial_inclusion is None:
+            partial_inclusion = self.partial_inclusion
+        newlabel = newlabel or self.label
+
+        if start_time is None:
+            start_time = self.time_ranges[0].start_time
+        if end_time is None:
+            end_time = self.time_ranges[-1].end_time
+
+        def keep_starting_on_or_before_end(item: tuple[TimeRange, Any]) -> bool:
+            time_range, _ = item
+            return time_range.start_time <= end_time
+
+        def keep_ending_on_or_after_start(item: tuple[TimeRange, Any]) -> bool:
+            time_range, _ = item
+            return time_range.end_time >= start_time
+
+        def keep_fully_overlapping(item: tuple[TimeRange, Any]) -> bool:
+            time_range, _ = item
+            if time_range.start_time < start_time:
+                return False
+            if time_range.end_time > end_time:
+                return False
+            return True
+
+        def replace_start_end_time(time_range: TimeRange) -> TimeRange:
+            start_time_ = max(time_range.start_time, start_time)
+            end_time_ = min(time_range.end_time, end_time)
+            return TimeRange(start_time_, end_time_)
```

Comment:
I refactored this part.

The keep_starting_on_.... functions are now reduced to two static functions (keep_overlapping and keep_fully_overlapping). replace_start_end_time is now also a static function.

The code has become more compact, but a bit harder to understand. It requires `partial`, because there is no other way (except for lambda's) to pass arguments to functions when using `filter` and `map`.

Reply:
I think zipping and unzipping is the best way to go here. Alternatively, we would need to filter and then keep score of which indices are kept, and then select the values with those indices.

Reply:
Thanks for looking into this. I actually don't think that compactness is more important than readability. I also made an attempt to [refactor ](https://github.com/EIT-ALIVE/eitprocessing/pull/168/files#r1625735887) it in a way that I think makes it more readable (and even more compact). See what you think.

Reply:
I reverted the commit, because it led to test failure.

I agree that compactness isn't always best. However, I really don't like `for` loops for filtering by appending accepted items to a new list. I would use either `filter()` or a list comprehension. 
I also don't like keeping a separate list of indices to then use that for selecting from a second list. 

I re-refactored the function using most of your ideas in d01912bf994db9aaf6f596bb9e5ee0d16d9903a0. What do you think?

Reply:
If it works then it's fine, and maybe not worth the time messing around with more.

That being said, to me, this solution still looks more complicated than I think it needs to be. 
- You are still creating a separate list of indices to select the values. In fact, you are creating 3 temporary lists (numbered_filtered_intervals, indices, filtered_intervals) versus 1 in my solution (keep_indices).
- The `if` statement in 144 doesn't really serve a purpose, apart from that zip fails on empty iterators. If numbered_filtered_intervals were a plain list, you could do away with the if and then an empty list would return empty lists as well. 
  - note: `if len(x)` is the same as `if x` (if x is not an iterator, `if len(x)` returns an error). Normally this triggers a ruff warning, not sure why that didn't happen here.
  - similarly, instead of `if self.values is not None`, we can use the property that we made for this `if not self.has_values` (also in concatenate)
- I agree that list comprehension/filter is normally cleaner and definitely more efficient than a "classical" for loop and appending items. But in this case, at least to me, the fact that multiple steps need to be taken (using indices on a separate list, trimming) makes the for loop/appending easier to read.
  - Actually, I think a list comprehension is only marginally more efficient than the for loop. Your first solution, using a filter, is definitely best in terms of efficiency, just makes reading it difficult.
  - if you really don't want to append, then maybe do everything in a single list comprehension to avoid having to avoid the if/else part below? Or is that even more confusing?:
```python
        iter_values = self.has_values or itertools.repeat(None)
        filtered_interval_value_pairs = [
            (iter_values[i], self._replace_start_end_time(interval, selection_start, selection_end))
            for i, interval in enumerate(self.intervals)
            if self._keep_overlapping(interval, selection_start, selection_end, partial_inclusion)
        ]
        # the return should have `values = values if self.has_values else None` or something similar
```
- I now see that the reason my suggestion failed is that I had passed `partial_inclusion` to `_replace_start_end_time` which it doesn't expect.